File: brownian_motion.py
import sys, math, time, datetime
import numpy as np
from matplotlib import pyplot as plt
from pylab import *

st = 100.0
runs = 1000
my = 0.045
iterations = 10000
variance = 0.3
delta_t = 1.0/iterations
counter = 0
barrier = 65

def plot_data(data, name):
    x = range(len(data))
    plot(x, data)
    xlabel('iteration')
    ylabel('stock price')
    title('About as simple as it gets, folks')
    grid(True)
    axhline(y=barrier, label="Barrier")
    savefig("test-{}.png".format(name))
    clf()
    #show()

start = datetime.datetime.now()
for times in range(runs):
    
    # Start from st every run
    st_next = st

    # Setup normalized random values
    e = np.random.randn(1,iterations)[0]
    
    broke = False
    data = []
    for dt_c in range(iterations):
        st_next = (st_next)*(math.e)**((my-0.5*(variance**2))*(delta_t)+(variance*e[dt_c]*math.sqrt(delta_t)))
        data.append(st_next)
                
        if st_next < barrier:
            broke = True
            break
    #plot_data(data, times)     
    if broke:
        counter += 1
end = datetime.datetime.now() - start
print "Broke", counter,"/",runs, "time(s)"
print "Time:", end
print "Rate:", counter*100.0/runs
print "Simulations/sec:", runs/end.total_seconds()

File: brownian_motion.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "randr.h"
#include <time.h>

#define RUNS       100000
#define ST         100
#define MY         0.45
#define ITERATIONS 1000
#define VARIANCE   0.3
#define DELTA_T    1.0/ITERATIONS
#define BARRIER    15
#define M_El	   2.718281828459045235360287471352662498

int main(void)
{
    clock_t begin, end;
    srand(time(NULL));
    double standard_dist[ITERATIONS];
    int broken = 0;
    double st_next = ST;
    begin = clock();
    for (int j = 0; j < RUNS; j++)
    {
        st_next = ST;
        randr_array(standard_dist, ITERATIONS);
        for (int i = 0; i < ITERATIONS; i++)
        {
            st_next = st_next * pow(M_El, ((MY-pow((0.5*VARIANCE),2))*(DELTA_T)+(standard_dist[i]*sqrt(DELTA_T))));
            if (st_next < BARRIER)
            {
                broken += 1;
                break;
            }
        }
        //printf("st_next: %f\n", st_next);
    }
    end = clock();
    printf("Broke: %d/%d\n", broken, RUNS);
    printf("Rate: %f%%\n", (broken*100.0)/RUNS);
    printf("Time spent: %f\n", (double)(end-begin)/CLOCKS_PER_SEC);
    printf("Simulation/sec: %f\n", (double)RUNS/((end-begin)/CLOCKS_PER_SEC));
}

